edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/ClientTokenCloudConnection.cs

```diff
@@ -0,0 +1,331 @@
+        /// Else, generate a notification of token being near expiry and return a task that
+        /// can be completed later.
+        /// Keep retrying till we get a usable token.
+        /// Note - Don't use this.Identity in this method, as it may not have been set yet!
+        /// </summary>
+        async Task<string> GetNewToken(string currentToken)
+        {
+            Events.GetNewToken(this.Identity.Id);
+            bool retrying = false;
+            string token = currentToken;
+            while (true)
+            {
+                // We have to catch UnauthorizedAccessException, because on IsTokenUsable, we call parse from
+                // Device Client and it throws if the token is expired.
+                if (IsTokenUsable(this.Identity.IotHubHostName, token))
+                {
+                    if (retrying)
+                    {
+                        Events.NewTokenObtained(this.Identity, token);
+                    }
+                    else
+                    {
+                        Events.UsingExistingToken(this.Identity.Id);
+                    }
+
+                    return token;
+                }
+                else
+                {
+                    Events.TokenNotUsable(this.Identity, token);
+                }
+
+                bool newTokenGetterCreated = false;
+                // No need to lock here as the lock is being held by the refresher.
+                TaskCompletionSource<string> tcs = this.tokenGetter
+                    .GetOrElse(
+                        () =>
+                        {
+                            Events.SafeCreateNewToken(this.Identity.Id);
+                            var taskCompletionSource = new TaskCompletionSource<string>();
+                            this.tokenGetter = Option.Some(taskCompletionSource);
+                            newTokenGetterCreated = true;
+                            return taskCompletionSource;
+                        });
+
+                // If a new tokenGetter was created, then invoke the connection status changed handler
+                if (newTokenGetterCreated)
+                {
+                    // If retrying, wait for some time.
+                    if (retrying)
+                    {
+                        await Task.Delay(TokenRetryWaitTime);
+                    }
+
+                    this.ConnectionStatusChangedHandler(this.Identity.Id, CloudConnectionStatus.TokenNearExpiry);
+                }
+
+                retrying = true;
+                // this.tokenGetter will be reset when this task returns.
+                token = await tcs.Task;
+            }
+        }
+
+        // Checks if the token expires too soon
+        static bool IsTokenUsable(string hostname, string token)
+        {
+            try
+            {
+                return TokenHelper.GetTokenExpiryTimeRemaining(hostname, token) > TokenExpiryBuffer;
+            }
+            catch (Exception e)
```

Comment:
I think catch should be moved to TokenHelper. It already has a catch for UnauthorizedAccessException